core/src/main/java/com/google/errorprone/bugpatterns/nullness/RedundantNullCheck.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2025 The Error Prone Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns.nullness;
+
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.matchers.Description.NO_MATCH;
+import static com.google.errorprone.util.ASTHelpers.hasImplicitType;
+import static com.google.errorprone.util.ASTHelpers.isConsideredFinal;
+import static javax.lang.model.type.TypeKind.TYPEVAR;
+
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.BinaryTreeMatcher;
+import com.google.errorprone.bugpatterns.nullness.NullnessUtils.NullCheck;
+import com.google.errorprone.dataflow.nullnesspropagation.Nullness;
+import com.google.errorprone.dataflow.nullnesspropagation.NullnessAnnotations;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+
+import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.LiteralTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+
+import java.util.Optional;
+
+import javax.lang.model.element.ElementKind;
+
+@BugPattern(
+    summary = "Explicit null check on a variable or method call that is not @Nullable within a @NullMarked scope.",
+    severity = WARNING)
+public class RedundantNullCheck extends BugChecker implements BinaryTreeMatcher {
+
+  @Override
+  public Description matchBinary(BinaryTree tree, VisitorState state) {
+    NullCheck nullCheck = NullnessUtils.getNullCheck(tree);
+    if (nullCheck == null) {
+      return NO_MATCH;
+    }
+
+    VarSymbol varSymbol = nullCheck.varSymbolButUsuallyPreferBareIdentifier();
+    if (varSymbol != null) {
+      if (!NullnessUtils.isInNullMarkedScope(varSymbol, state)) {
```

Comment:
I hadn't thought about this before, even when I commented on the docs on the subject, but: Does checking for null-marked buy us anything? Any information that we get about a variable is either from the method whose result it's initialized to or from "language knowledge" (constructors, literals). So we're ([as usual](https://jspecify.dev/docs/api/org/jspecify/annotations/Nullable.html#applicability)) not relying on the annotations for the variable's "top-level" type.

(This would change if we were to start handling _fields_, as discussed in [a thread below](https://github.com/google/error-prone/pull/5121/files#r2373414895). That's all the more reason to _not_ do that yet :))